KIC Get Started (#1068)
Kong/developer.konghq.com: app/_how-tos/kic-get-started-services-routes.md

---
title: Services and Routes
description: |
  Use the a HTTPRoute or Ingress to configure a Service and a Route.
content_type: how_to

permalink: /kubernetes-ingress-controller/get-started/services-and-routes/
breadcrumbs:
  - /kubernetes-ingress-controller/
  - index: kubernetes-ingress-controller
    section: Get Started

series:
  id: kic-get-started
  position: 2

tldr:

>   q: How do I create a Service and a Route with {{ site.kic_product_name }}

Reviewer: ```suggestion
  q: How do I create a Service and a Route with {{ site.kic_product_name }}?
```

  a: |
    Create a HTTPRoute that points to a Kubernetes service in your cluster.

products:
  - kic

works_on:
  - on-prem
  - konnect

prereqs:
  skip_product: true
---

> ## How Kubernetes resources map to Kong Entities

Reviewer: ```suggestion
## How Kubernetes resources map to {{site.base_gateway}} entities
```

> A Service inside Kubernetes is a way to abstract an application that is running on a set of Pods. This maps to two objects in Kong: Service and Upstream.

Reviewer: ```suggestion
A Service inside Kubernetes is a way to abstract an application that is running on a set of Pods. This maps to two entities in {{site.base_gateway}}: Service and Upstream.
```
For consistency, let's use "entities" instead of "objects".

> The service object in Kong holds the information of the protocol to use to talk to the upstream service and various other protocol specific settings. The Upstream object defines load-balancing and health-checking behavior.

Reviewer: ```suggestion
The Service entity in {{site.base_gateway}} holds the protocol information needed to talk to the upstream service and various other protocol-specific settings. The Upstream object defines load balancing and health-checking behavior.
```

<!--vale off-->
{% mermaid %}
flowchart LR
    H(Request traffic)
    subgraph Pods
        direction LR
        E(Target)
        F(Target)
        G(Target)
    end

    subgraph Kubernetes Service
        direction TB
        C(Service)
        D(Upstream)
    end
    
    subgraph Ingress / HTTPRoute
        direction LR
        A(Route)
        B(Route)
    end

    A --> C
    B --> C
    C --> D
    D --> E
    D --> F
    D --> G
    H --> A

    linkStyle 6 stroke:#b6d7a8
{% endmermaid %}
<!--vale on-->

Routes are configured using Gateway API or Ingress resources, such as `HTTPRoute`, `TCPRoute`, `GRPCRoute`, `Ingress` and more.

## Deploy an echo service

> In this guide, you will deploy an `echo` service which returns information about the Kubernetes cluster and route traffic to the service.

Reviewer: ```suggestion
Let's start by deploying an `echo` service which returns information about the Kubernetes cluster and routes traffic to the service.
```
This should be a preamble for the command, and it doesn't currently read that way.

```bash
kubectl apply -f {{ site.links.web }}/manifests/kic/echo-service.yaml -n kong
```

> ## Create a HTTPRoute / Ingress

Reviewer: ```suggestion
## Create an HTTPRoute / Ingress
```
American English :)

Author: I'm not sure on this one. I've always used `an` when the H is silent e.g. `an hour`, but `a` when it's pronounced e.g. `HTTP`

Author: Alright, I spent some time thinking. It sounded wrong at first, but I think I got used to it

> 

Reviewer: This needs a description. Why are we creating this route? What is it going to do?

{% include /k8s/httproute.md release=page.release path='/echo' name='echo' service='echo' port='1027' skip_host=true %}

## Validate your configuration

> Once the resource has been reconciled, you'll be able to call the `/echo` endpoint and {{ site.base_gateway }} will route the request to the `echo` service.

Reviewer: ```suggestion
Once the resource has been reconciled, you can call the `/echo` endpoint and {{ site.base_gateway }} will route the request to the `echo` service:
```

{% validation request-check %}
url: /echo
status_code: 200

> on_prem_url: $PROXY_IP

Reviewer: Not working for me (and PROXY_IP is set):
```
➜  ~ echo "
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: echo
  namespace: kong
  annotations:
    konghq.com/strip-path: 'true'
spec:
  ingressClassName: kong
  rules:
  - http:
      paths:
      - path: /echo
        pathType: ImplementationSpecific
        backend:
          service:
            name: echo
            port:
              number: 1027
" | kubectl apply -f -

ingress.networking.k8s.io/echo created
➜  ~ curl $PROXY_IP/echo             

{
  "message":"no Route matched with those values",
  "request_id":"ce62ce532171d91afd215521841b5439"
}%    
```

Reviewer: Redid the whole thing from the beginning, no problem here now. Probably missed a step somewhere.

> on_prem_url: $PROXY_IP

Author: I'll add this to the Weds sync. The introduction of the `series` concept breaks EPPO somewhat. One to discuss if the tradeoffs are worth it before rolling out more widely or removing series
